paimon-core/src/main/java/org/apache/paimon/schema/SchemaMergingUtils.java

```diff
@@ -0,0 +1,196 @@
+                                                    baseField.id(),
+                                                    baseField.name(),
+                                                    updatedDataType,
+                                                    baseField.description());
+                                        } else {
+                                            return baseField;
+                                        }
+                                    })
+                            .collect(Collectors.toList());
+
+            Map<String, DataField> baseFieldMap =
+                    baseFields.stream()
+                            .collect(Collectors.toMap(DataField::name, Function.identity()));
+            List<DataField> newFields =
+                    updateFields.stream()
+                            .filter(field -> !baseFieldMap.containsKey(field.name()))
+                            .map(field -> assignIdForNewField(field, highestFieldId))
+                            .collect(Collectors.toList());
+
+            updatedFields.addAll(newFields);
+            return new RowType(base.isNullable(), updatedFields);
+        } else if (base instanceof MapType && update instanceof MapType) {
+            return new MapType(
+                    base.isNullable(),
+                    merge(
+                            ((MapType) base).getKeyType(),
+                            ((MapType) update).getKeyType(),
+                            highestFieldId),
+                    merge(
+                            ((MapType) base).getValueType(),
+                            ((MapType) update).getValueType(),
+                            highestFieldId));
+        } else if (base instanceof ArrayType && update instanceof ArrayType) {
+            return new ArrayType(
+                    base.isNullable(),
+                    merge(
+                            ((ArrayType) base).getElementType(),
+                            ((ArrayType) update).getElementType(),
+                            highestFieldId));
+        } else if (base instanceof MultisetType && update instanceof MultisetType) {
+            return new MultisetType(
+                    base.isNullable(),
+                    merge(
+                            ((MultisetType) base).getElementType(),
+                            ((MultisetType) update).getElementType(),
+                            highestFieldId));
+        } else if (base instanceof DecimalType && update instanceof DecimalType) {
+            if (base.equals(update)) {
+                return base0;
+            } else {
+                throw new UnsupportedOperationException(
+                        String.format(
+                                "Failed to merge decimal types with different precision or scale: %s and %s",
+                                base, update));
+            }
+        } else if (DataTypeCasts.supportsImplicitCast(base, update)) {
```

Comment:
Currently, we have using `supportsExplicitCast` to manually casting (From DDL alter table).
Do you think we should use `supportsExplicitCast`? Or provide an option to let user to choose?

Reply:
that needs an option if we want to support Explicit-Casting.